### ISSUE TITLE
Duplicate alerts

### ISSUE THREAD

**Issue by troyraen** [mwvgroup/Pitt-Google-Broker#173]
This issue documents a current problem the broker is experiencing with duplicate alerts, and starts the more general discussion of how the broker should handle duplicates.

Currently, about twice as many alerts end up in our `ztf-alerts` stream (published by our consumer) as are produced by ZTF each night. I have spent some time trying to track down where the duplicates are coming from. I think they're actually coming from ZTF, but I can't say for sure yet. 

I know that ZTF is currently having problems with duplicate alerts on their end (sometimes enough to fill up all their disk space), but they haven't been able to tell me specifically whether they're actually sending duplicates to us. I've also asked them whether they can tell if we're simply requesting/pulling the stream from them twice, but I don't have an answer on that yet either. Last night we actually received 4 times the number of alerts produced by ZTF. Half those arrived within a 45 minute window at about 7 am Pacific... very reminiscent of the stream replays that I used to see from ZTF every few days (which I "fixed" by shutting the broker down a little earlier each morning).

I've looked at various aspects of our consumer, but I haven't found any evidence that it is either requesting/pulling in the stream twice or producing duplicates internally. But I found no confirmation that it's **_not_** doing one of those things either.

The one thing I do feel confident about is that we do **_not_** have some other resource hiding somewhere in Google Cloud that is publishing to the `ztf-alerts` stream -- both because I've looked fairly thoroughly and because alerts do not end up in that stream unless our `ztf-consumer` VM is actually running.

Regardless of where they're coming from, the duplicates can be expensive for us. This is especially true when a sudden influx causes a backup in the modules that load to BigQuery, because we end up paying for stupidly-long runtimes of Cloud Functions (see https://github.com/mwvgroup/Pitt-Google-Broker/issues/96#issuecomment-929401416).

The bigger, more general question of how to make the broker handle duplicates more gracefully is also tricky. Duplicates can come from several different places, and the de-duplication approach is different for each. Worse, depending on the specifics, there may not even **_be_** a straightforward approach. To illustrate:  How would the broker even know if a given message was a duplicate? It can't check the entire data packet. It would have to track some indicator, like a Kafka offset or a Pub/Sub message ID. But if (for example) our consumer is just publishing the same message twice, those messages will have **_different_** IDs even though they contain the same data packet.

**Comment by wmwv** [comment]
Each stream must generate some unique AlertID.  How expensive is it to filter on "streamName+alertID" for de-duplication?

**Comment by troyraen** [comment]
Documenting for future readers (@wmwv and I have already discussed offline).

TLDR is that I found a solution that takes advantage of Cloud Storage to do this with $0 in extra cost and less than 20 lines of code changes (see #183). I'm thrilled that the fix is so simple as this problem was starting to look like a nightmare...

> How expensive is it to filter on "streamName+alertID" for de-duplication?

There is not even a _mechanism_ to do this in either the Kafka connector that runs our consumer or in Pub/Sub itself, and writing our own solution is daunting. The biggest problem is that this task cannot really be parallelized. There are different ways to implement it, but essentially one has to keep a list of published messages and check against it for duplicates before publishing more. To make that work, the list really has to be a 1-out-1-in situation, where "out" = "check the list for a given message's AlertID" and "in" = "publish the message (assuming no duplicate) and register the AlertID in the list". 

Being massively parallel is an explicit and fundamental design principle of Pub/Sub, so I suppose it's not surprising that it (they/developers) doesn't want to touch this problem of duplicates in data streams. The Kafka connector is maintained by Pub/Sub developers, who probably don't want to touch it there for the same reasons. 

The non-parallel requirement also means that we can't solve this ourselves with a Cloud Function or Cloud Run, since those also run in parallel in a very fundamental way. Instead we'd have to run it on a VM... and now I'm having nightmares about the Metadata Collector (see #171) and Dataflow (which we got rid of in #94, though it seems I didn't put a good description anywhere in the repo about the headaches it caused). 

Google's recommended solution is to use Dataflow. No thanks.

Then I realized that Cloud Storage can do something similar. It has a mechanism to prevent a file from uploading if an object with the same name already exists in the bucket. #183 takes advantage of this mechanism to implement the fix.

**Comment by troyraen** [comment]
Posting some screenshots of October 1, 2022 (a relatively random, fairly normal night) to document the problem a little better, and the performance of the Cloud Storage solution.

**Notes**

Screenshots with:
- black background come from [ZTF's dashboard](https://monitor.alerts.ztf.uw.edu/d/XAs4s7Giz/ztf-alert-distribution-system-status) (more on these below*)
- white background come from [our dashboard](https://console.cloud.google.com/monitoring/dashboards/builder/broker-instance-ztf-False).

Our streams:
- `ztf-alerts_raw` is published by our Consumer
- `ztf-alerts` stream is published by our Cloud Storage module after de-duplicating the `ztf-alerts_raw` stream

### Total alerts (cumulative for the night)

- 217,363 reported by ZTF in at least one place:

<img width="151" alt="Screen Shot 2022-10-22 at 12 50 39 AM" src="https://user-images.githubusercontent.com/29547463/197327609-6eb57eb5-d792-4181-9a15-0cdc1aabc0dc.png">

- 442,070 in our `ztf-alerts_raw` stream
- 217,360 in our `ztf-alerts` stream (only 5 sig figs):

<img width="1431" alt="Screen Shot 2022-10-21 at 11 05 36 PM" src="https://user-images.githubusercontent.com/29547463/197328105-59c3a387-dec4-4424-afbd-718e68583032.png">

**So it looks like the de-duplication is working very well. I haven't checked details, but it seems like we're actually keeping all the originals and dropping all the duplicates.**

FWIW, ZTF's version of this figure (cumulative alerts, above) makes no sense. There are two lines that supposedly represent the same data, but neither correspond to any nightly total quoted above:

<img width="1364" alt="Screen Shot 2022-10-21 at 11 05 12 PM" src="https://user-images.githubusercontent.com/29547463/197328318-47f5abab-923c-4890-b508-91220f629d2f.png">

### Alert rates

- ZTF has 8 peaks above 6,000 alerts / min:

<img width="1365" alt="Screen Shot 2022-10-21 at 11 13 18 PM" src="https://user-images.githubusercontent.com/29547463/197328966-3bf547e3-7a91-4cc1-93ba-cd0d082acfb6.png">

Both of our streams peak at about the same times, 
- `ztf-alerts_raw` peaks above 6,000 / min (100 / sec)
- `ztf-alerts` peaks are roughly half as big:

<img width="1436" alt="Screen Shot 2022-10-21 at 11 21 44 PM" src="https://user-images.githubusercontent.com/29547463/197331717-0749c1a7-4177-4b1d-ab92-76e8af062700.png">

---

*More on ZTF's screenshots:

These show "received" and "incoming". As I understand it, these show alerts that are incoming to the ZTF Alert Distribution System (ZADS), from IPAC. The alerts are produced at IPAC (in Pasadena) and then sent to the ZADS, which runs their Kafka broker (which sends the alerts to us) and lives on an AWS cluster in San Francisco. 

This seems to suggest that the duplicates are coming from IPAC, but that seems contrary to the fact the we are seeing similar duplicates in the ELAsTiCC stream. (ELAsTiCC uses ZTF's Kafka broker to distribute alerts, but I'm fairly certain they are sent to the ZADS from NERSC and never go through IPAC.)

**Comment by JulienPeloton** [comment]
Hi @troyraen -- sorry to sneak in here :D

In Fink, I received duplicate ZTF alerts for a long time. Inspecting my logs, and performing simulations, I could tell it was not coming from my systems. I talked to people at IPAC but they couldn't locate the problem either. Recently, we switch ZTF producer (we were polling from the sec_broker server -- public2.alerts -- which is a mirror of their primary producer on public.alerts), and magically duplicates disappeared. Could it be that you are also polling from public2.alerts?

**Comment by troyraen** [comment]
@JulienPeloton Thank you! Yes, we very likely are connecting to public2. I'll have to get in and look / change it after my workday. Thanks for the heads up.
